keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1441 @@
+
+  // If specified (non-zero value), the Plugin MUST NOT return more
+  // entries than this number in the response. If the actual number of
+  // entries is more than this number, the Plugin MUST set `next_token`
+  // in the response which can be used to get the next page of entries
+  // in the subsequent `ListSnapshotDeltas` call. This field is
+  // OPTIONAL. If not specified (zero value), it means there is no
+  // restriction on the number of entries that can be returned.
+  // The value of this field MUST NOT be negative.
+  int32 max_entries = 4;
+
+  // A token to specify where to start paginating. Set this field to
+  // `next_token` returned by a previous `ListSnapshotDeltas` call to
+  // get the next page of entries. This field is OPTIONAL.
+  // An empty string is equal to an unspecified field value.
+  string starting_token = 5;
+}
+
+message ListSnapshotDeltasResponse {
+  option (alpha_message) = true;
+
+  // The volume size in bytes. This field is OPTIONAL.
+  uint64 volume_size_bytes = 1;
+
+  // This token allows you to get the next page of entries for
+  // `ListSnapshotDeltas` request. If the number of entries is larger
+  // than `max_entries`, use the `next_token` as a value for the
+  // `starting_token` field in the next `ListSnapshotDeltas` request.
+  // This field is OPTIONAL.
+  // An empty string is equal to an unspecified field value.
+  string next_token = 2;
+
+  // Changed block deltas between the source and target snapshots. An
+  // empty list means there is no difference between the two. Leave
+  // unspecified if the volume isn't of block type. This field is
+  // OPTIONAL.
+  repeated BlockSnapshotChangedBlock changed_blocks = 3;
+}
+
+message BlockSnapshotChangedBlock {
+  option (alpha_message) = true;
+
+  // The block logical offset on the volume. This field is REQUIRED.
+  uint64 offset = 1;
+
+  // The size of the block in bytes. This field is REQUIRED.
+  uint64 block_size_bytes = 2;
+
+  // The token and other information needed to retrieve the actual
+  // data block at the given offset. If the provider doesn't support
+  // token-based data blocks retrieval, this should be left
+  // unspecified. This field is OPTIONAL.
+  BlockSnapshotChangedBlockToken token = 3;
+}
+
+message BlockSnapshotChangedBlockToken {
+  option (alpha_message) = true;
+
+  // The token to use to retrieve the actual data block at the given
+  // offset. This field is REQUIRED.
+  string token = 1;
+
+  // Timestamp when the token is issued. This field is REQUIRED.
+  .google.protobuf.Timestamp issuance_time = 2;
+
+  // The TTL of the token in seconds. The expiry time is calculated by
+  // adding the time of issuance with this value. This field is
+  // REQUIRED.
+  int32 ttl_seconds = 3;
+}
+```
+
+#### DriverDiscovery Resource
+
+The section describes the specification of `DriverDiscovery` resource:
+
+```go
+// DriverDiscovery represents a DriverDiscovery resource.
+type DriverDiscovery struct {
```

Comment:
(starting a separate thread), what is the reason it's not part of CSIDriver object? It should be mentioned somewhere in the KEP.

Reply:
I forgot to add that to the "Alternatives" section - my bad. I was never given the full reason why we can't put this in the CSIDriver, but from what I gathered, it's because CSIDriver is used mainly for properties needed by the kubelet, at the node-level. Does that sound right to you?

Reply:
Yes, I thougt about it a bit more and the main reason for a new object for me is that snapshots are out of tree and provided by CRDs. Having parts of snapshot API in-tree in CSIDriver is odd.